File: python/numerical_stress_tests.py
#!/usr/bin/env python3
"""Independent numerical stress tests for tensor conventions and integrals."""

from __future__ import annotations

from itertools import permutations

import numpy as np


def build_tensor(values: np.ndarray) -> np.ndarray:
    if values.shape != (7,):
        raise ValueError("Expected seven tensor parameters.")
    a, b, c, d, e, f, g = values
    seeds = {
        (0, 0, 0): a,
        (0, 0, 1): b,
        (0, 0, 2): c,
        (0, 1, 1): d,
        (0, 1, 2): e,
        (1, 1, 1): f,
        (1, 1, 2): g,
        (0, 2, 2): -a - d,
        (1, 2, 2): -b - f,
        (2, 2, 2): -c - g,
    }
    tensor = np.zeros((3, 3, 3), dtype=float)
    for key, value in seeds.items():
        for p in set(permutations(key)):
            tensor[p] = value
    return tensor


def evaluate_quantities(tensor: np.ndarray, u: np.ndarray) -> tuple[float, float, float]:
    harmonic = np.einsum("ijk,i,j,k->", tensor, u, u, u)
    matrix_tu = np.einsum("ijk,i->jk", tensor, u)
    vector_tuu = np.einsum("ijk,i,j->k", tensor, u, u)
    projection = np.eye(3) - np.outer(u, u)
    curvature = projection @ (6.0 * matrix_tu) @ projection - 2.0 * harmonic * projection
    norm_sq = float(np.trace(curvature @ curvature))
    q_projected = float(2.0 * (np.trace(curvature) ** 2 - norm_sq))
    q_compact = float(
        64.0 * harmonic**2
        + 144.0 * np.dot(vector_tuu, vector_tuu)
        - 72.0 * np.trace(matrix_tu @ matrix_tu)
    )
    return norm_sq, q_projected, q_compact


def gauss_legendre_sphere(n_z: int = 24, n_phi: int = 64) -> tuple[np.ndarray, np.ndarray]:
    z, w_z = np.polynomial.legendre.leggauss(n_z)
    phi = 2.0 * np.pi * np.arange(n_phi) / n_phi
    points = []
    weights = []
    for zi, wi in zip(z, w_z, strict=True):
        radius = np.sqrt(max(0.0, 1.0 - zi * zi))
        for angle in phi:
            points.append([radius * np.cos(angle), radius * np.sin(angle), zi])
            weights.append(wi * 2.0 * np.pi / n_phi)
    return np.asarray(points), np.asarray(weights)


def tensor_invariants(tensor: np.ndarray) -> tuple[float, float]:
    tau = float(np.einsum("ijk,ijk->", tensor, tensor))
    b_matrix = np.einsum("ikl,jkl->ij", tensor, tensor)
    b0 = b_matrix - np.eye(3) * tau / 3.0
    b0_sq = float(np.trace(b0 @ b0))
    return tau, b0_sq


def random_rotation(rng: np.random.Generator) -> np.ndarray:
    q, _ = np.linalg.qr(rng.normal(size=(3, 3)))
    if np.linalg.det(q) < 0:
        q[:, 0] *= -1
    return q


def rotate_tensor(tensor: np.ndarray, rotation: np.ndarray) -> np.ndarray:
    return np.einsum("ia,jb,kc,abc->ijk", rotation, rotation, rotation, tensor)


def main() -> None:
    rng = np.random.default_rng(20260902)
    sphere_points, sphere_weights = gauss_legendre_sphere()

    max_pointwise_error = 0.0
    max_norm_relative_error = 0.0
    max_q_relative_error = 0.0
    max_rotation_error = 0.0

    for _ in range(12):
        tensor = build_tensor(rng.normal(size=7))
        tau, b0_sq = tensor_invariants(tensor)
        predicted_norm = np.pi * (176.0 / 7.0) * tau
        predicted_q_sq = np.pi / 1001.0 * (555264.0 * tau**2 - 2265600.0 * b0_sq)

        integrated_norm = 0.0
        integrated_q_sq = 0.0
        for point, weight in zip(sphere_points, sphere_weights, strict=True):
            norm_sq, q_direct, q_compact = evaluate_quantities(tensor, point)
            max_pointwise_error = max(max_pointwise_error, abs(q_direct - q_compact))
            integrated_norm += weight * norm_sq
            integrated_q_sq += weight * q_direct**2

        max_norm_relative_error = max(
            max_norm_relative_error,
            abs(integrated_norm - predicted_norm) / max(1.0, abs(predicted_norm)),
        )
        max_q_relative_error = max(
            max_q_relative_error,
            abs(integrated_q_sq - predicted_q_sq) / max(1.0, abs(predicted_q_sq)),
        )

        rotation = random_rotation(rng)
        rotated = rotate_tensor(tensor, rotation)
        tau_rotated, b0_sq_rotated = tensor_invariants(rotated)
        max_rotation_error = max(
            max_rotation_error,
            abs(tau - tau_rotated),
            abs(b0_sq - b0_sq_rotated),
        )

    if max_pointwise_error > 2e-10:
        raise AssertionError(f"Pointwise determinant mismatch: {max_pointwise_error}")
    if max_norm_relative_error > 2e-12:
        raise AssertionError(f"Norm integral mismatch: {max_norm_relative_error}")
    if max_q_relative_error > 2e-11:
        raise AssertionError(f"q^2 integral mismatch: {max_q_relative_error}")
    if max_rotation_error > 2e-11:
        raise AssertionError(f"Rotation-invariance mismatch: {max_rotation_error}")

    print("NUMERICAL STRESS TESTS: PASS")
    print(f"maximum pointwise determinant error = {max_pointwise_error:.3e}")
    print(f"maximum norm relative error          = {max_norm_relative_error:.3e}")
    print(f"maximum q^2 relative error           = {max_q_relative_error:.3e}")
    print(f"maximum invariant rotation error     = {max_rotation_error:.3e}")


if __name__ == "__main__":
    main()

File: R/numerical_stress_tests.R
#!/usr/bin/env Rscript
# Independent base-R numerical checks of the degree-3 tensor conventions.
# This is a numerical cross-check; the Python/SymPy scripts perform the exact proof.

build_tensor <- function(values) {
  stopifnot(length(values) == 7)
  a <- values[1]; b <- values[2]; c <- values[3]; d <- values[4]
  e <- values[5]; f <- values[6]; g <- values[7]
  T <- array(0, dim = c(3, 3, 3))

  set_sym <- function(indices, value) {
    perms <- unique(rbind(
      indices,
      indices[c(1, 3, 2)],
      indices[c(2, 1, 3)],
      indices[c(2, 3, 1)],
      indices[c(3, 1, 2)],
      indices[c(3, 2, 1)]
    ))
    for (row in seq_len(nrow(perms))) {
      T[perms[row, 1], perms[row, 2], perms[row, 3]] <<- value
    }
  }

  set_sym(c(1, 1, 1), a)
  set_sym(c(1, 1, 2), b)
  set_sym(c(1, 1, 3), c)
  set_sym(c(1, 2, 2), d)
  set_sym(c(1, 2, 3), e)
  set_sym(c(2, 2, 2), f)
  set_sym(c(2, 2, 3), g)
  set_sym(c(1, 3, 3), -a - d)
  set_sym(c(2, 3, 3), -b - f)
  set_sym(c(3, 3, 3), -c - g)
  T
}

contract_Tuuu <- function(T, u) {
  total <- 0
  for (i in 1:3) for (j in 1:3) for (k in 1:3) {
    total <- total + T[i, j, k] * u[i] * u[j] * u[k]
  }
  total
}

matrix_Tu <- function(T, u) {
  M <- matrix(0, 3, 3)
  for (j in 1:3) for (k in 1:3) for (i in 1:3) {
    M[j, k] <- M[j, k] + T[i, j, k] * u[i]
  }
  M
}

vector_Tuu <- function(T, u) {
  v <- numeric(3)
  for (k in 1:3) for (i in 1:3) for (j in 1:3) {
    v[k] <- v[k] + T[i, j, k] * u[i] * u[j]
  }
  v
}

evaluate_quantities <- function(T, u) {
  Y <- contract_Tuuu(T, u)
  M <- matrix_Tu(T, u)
  v <- vector_Tuu(T, u)
  P <- diag(3) - tcrossprod(u)
  C <- P %*% (6 * M) %*% P - 2 * Y * P
  norm_sq <- sum(diag(C %*% C))
  q_direct <- 2 * (sum(diag(C))^2 - norm_sq)
  q_compact <- 64 * Y^2 + 144 * sum(v^2) - 72 * sum(diag(M %*% M))
  c(norm_sq = norm_sq, q_direct = q_direct, q_compact = q_compact)
}

gauss_legendre <- function(n) {
  i <- seq_len(n - 1)
  beta <- i / sqrt(4 * i^2 - 1)
  J <- matrix(0, n, n)
  J[cbind(i, i + 1)] <- beta
  J[cbind(i + 1, i)] <- beta
  eig <- eigen(J, symmetric = TRUE)
  ord <- order(eig$values)
  list(nodes = eig$values[ord], weights = 2 * eig$vectors[1, ord]^2)
}

sphere_grid <- function(n_z = 24, n_phi = 64) {
  gl <- gauss_legendre(n_z)
  phi <- 2 * pi * (0:(n_phi - 1)) / n_phi
  points <- matrix(0, n_z * n_phi, 3)
  weights <- numeric(n_z * n_phi)
  row <- 1
  for (iz in seq_len(n_z)) {
    z <- gl$nodes[iz]
    r <- sqrt(max(0, 1 - z^2))
    for (angle in phi) {
      points[row, ] <- c(r * cos(angle), r * sin(angle), z)
      weights[row] <- gl$weights[iz] * 2 * pi / n_phi
      row <- row + 1
    }
  }
  list(points = points, weights = weights)
}

tensor_invariants <- function(T) {
  tau <- sum(T^2)
  B <- matrix(0, 3, 3)
  for (i in 1:3) for (j in 1:3) for (k in 1:3) for (ell in 1:3) {
    B[i, j] <- B[i, j] + T[i, k, ell] * T[j, k, ell]
  }
  B0 <- B - diag(3) * tau / 3
  c(tau = tau, b0_sq = sum(diag(B0 %*% B0)))
}

set.seed(20260902)
grid <- sphere_grid()
max_pointwise <- 0
max_norm_rel <- 0
max_q_rel <- 0

for (trial in 1:12) {
  T <- build_tensor(rnorm(7))
  inv <- tensor_invariants(T)
  predicted_norm <- pi * (176 / 7) * inv["tau"]
  predicted_q_sq <- pi / 1001 * (555264 * inv["tau"]^2 - 2265600 * inv["b0_sq"])

  norm_integral <- 0
  q_integral <- 0
  for (row in seq_len(nrow(grid$points))) {
    values <- evaluate_quantities(T, grid$points[row, ])
    max_pointwise <- max(max_pointwise, abs(values["q_direct"] - values["q_compact"]))
    norm_integral <- norm_integral + grid$weights[row] * values["norm_sq"]
    q_integral <- q_integral + grid$weights[row] * values["q_direct"]^2
  }
  max_norm_rel <- max(max_norm_rel, abs(norm_integral - predicted_norm) / max(1, abs(predicted_norm)))
  max_q_rel <- max(max_q_rel, abs(q_integral - predicted_q_sq) / max(1, abs(predicted_q_sq)))
}

stopifnot(max_pointwise < 5e-9)
stopifnot(max_norm_rel < 5e-11)
stopifnot(max_q_rel < 5e-10)

ratio <- 15183 / 17303
candidate <- pi / 1914 * (259 - 27 * sqrt(ratio))
nishioka <- 4 * pi / 33
meissner <- pi * (2 / 3 - sqrt(3) / 4 * acos(1 / 3))
stopifnot(nishioka < candidate, candidate < meissner)

cat("R NUMERICAL STRESS TESTS: PASS\n")
cat(sprintf("maximum pointwise determinant error = %.3e\n", max_pointwise))
cat(sprintf("maximum norm relative error          = %.3e\n", max_norm_rel))
cat(sprintf("maximum q^2 relative error           = %.3e\n", max_q_rel))
cat(sprintf("candidate coefficient                = %.15f\n", candidate))
